### ISSUE TITLE
Client Authentication Policy "Not required when using PKCE" blocking the refresh token grant

### ISSUE THREAD

**Issue by TanguyGiton** [FusionAuth/fusionauth-issues#1653]
### Description 

When we choose the option "Not required when using PKCE" for Client Authentication, it's working well with the grant "Authorization Code", but can't refresh the token with the Refresh Token Grant. We can't use PKCE for this grant.

### Affects versions
1.32.1

### Steps to reproduce

1. Go to an application
2. Select "Not required when using PKCE" option in Client Authentication
3. Enable "Authorization Code" and "Refresh Token" grants
4. Activate Generate Refresh Token
5. Get a token with PKCE, "Authorization Code" grant and "offline_access" scope
6. Refresh the token
7. Error "Client authentication missing as Basic Authorization header or credentials in the body (or some combination of them)."

### Expected behavior
We should use the refresh token grant without PKCE

### Community guidelines
All issues filed in this repository must abide by the [FusionAuth community guidelines](https://fusionauth.io/community/forum/topic/1000/code-of-conduct).

**Comment by robotdan** [comment]
This seems to be working as designed. PKCE is not applicable to the Refresh Grant, so you then must provide client authentication using the client credentials (client_id and client_secret). 

Are you expecting the Refresh Grant to work without any client authentication?

**Comment by TanguyGiton** [comment]
Yes, as PKCE is not application the Refresh Grant, I think client authentication should not be required when choosing option "Not required when using PKCE".

**Comment by robotdan** [comment]
Hmm...  will need to think more about this. I would hesitate to turn off client authentication for the refresh grant implicitly based upon this configuration. This configuration only applies to grants that can utilize PKCE, otherwise client authentication is always required. 

To me, this is working as designed. 

I think if this is something we would want to support, we'd have to add an additional configuration to indicate if client authentication is required for a refresh grant or not. 

What is the use case that you have that you do not want to use client authentication? In most cases you will want to refresh a JWT from some sort of back end that is then able to write an HTTP only secure cookie. If you do this from within a user-agent, you cannot write a secure HTTP only cookie which makes the JWT vulnerable.

**Comment by glen-84** [comment]
@robotdan,

I stumbled upon this issue while evaluating FusionAuth. I'm pretty sure that a client secret should _not_ be required when using refresh tokens. The whole point of PKCE is to enable public clients to use OAuth without having to store secrets, which it cannot do in SPA or mobile applications.

> But what about a Refresh Token flow? When using a refresh token, confidential clients also have to authenticate. Public clients, such as browser-based applications, do not authenticate during the Refresh Token flow. **So in a typical frontend application, refresh tokens issued to frontend web applications are bearer tokens**.
>
> In practice, this means that if an attacker manages to steal a refresh token from a frontend application, they can use that token in a Refresh Token flow. To counter such attacks, the OAuth 2.0 specifications mandate that browser-based applications apply a security measure known as **refresh token rotation**.
>
> – https://www.pingidentity.com/en/resources/blog/posts/2021/refresh-token-rotation-spa.html

I also evaluated Auth0 recently, and their SPA SDK handled this all (including PKCE and refresh token rotation) without using a client secret.

Just to confirm, does FusionAuth support refresh token rotation via the `Refresh Token usage` setting set to `One time use`?

This may be a bit of a blocker for us, so hopefully this can be fixed soon.

**Comment by robotdan** [comment]
According to https://www.rfc-editor.org/rfc/rfc6749#section-6

> If the client type is confidential or
   the client was issued client credentials (or assigned other
   authentication requirements), the client MUST authenticate with the
   authorization server as described in [Section 3.2.1](https://www.rfc-editor.org/rfc/rfc6749#section-3.2.1).
> ...   
>    The authorization server MUST:
>
>   o  require client authentication for confidential clients or for any
      client that was issued client credentials (or with other
      authentication requirements),
>
>   o  authenticate the client if client authentication is included and
      ensure that the refresh token was issued to the authenticated
      client, and


I don't know that we should break spec here. It is possible that Auth0 has allowed you to configure a public client, and in that case authentication is not required. We have a similar configuration, you can set "Client Authentication: Not required" which then disables this requirement. 

Not sure I understand how this is a blocker, can you further explain your use case?

**Comment by glen-84** [comment]
I would not want you to break spec! 🙂

I guess I got a bit confused by the OP, and forgot that it could be disabled entirely. I apologize for that.

So "Not required when using PKCE" is only for confidential clients then, and the OP should instead use "Not required"?

**Comment by TanguyGiton** [comment]
That's what I did in this case, I used "Client Authentification: Not required".

For me, the option "Client Authentification: Not required when using PKCE" allow us to have both SPA and backend authenticating on the same FusionAuth application (with shared roles).

On the SPA, I want to use refresh token without authentication because the JWT expire after 1 hour, so I want to refresh it in the background without logging out the user.

On the SPA, the refresh token is only stored in memory (no cookie or local storage) and it is configured on "Sliding window" and "One time use".

**Comment by huysentruitw** [comment]
Stumbled across the same issue. We can't use the `refresh_token` grant from front-end, unless we set `Client Authentication` to  `Not required`. Which is a bummer, as we want to use PKCE for `authorization_code` grant. To bad both grants are configured with a single options.
